### ISSUE TITLE
Code Server's Process Still Running After Stopping the maven launcher ( CTRL + C ) 

### ISSUE THREAD

**Issue by zouabimourad** [tbroyer/gwt-maven-plugin#110]
Hi,

When i launch `mvn gwt:codesever` 3 process are started ( maven jvm , cmd and codesever jvm ) 

When I Stop mvn .. the 2 first process are stopped but the codeserver jvm is still running

Regards.

**Comment by zouabimourad** [comment]
Same thing when i stop gwt compilation .. the leaf java process ( gwt compiler process )  is still running ...

**Comment by zouabimourad** [comment]
I have the same problem with the 1.8.2 version of the  legacy plugin but the 1.8.1 is ok 

I will try to fix it.

**Comment by zouabimourad** [comment]
It works fine with the dependency override.  There is no a `cmd` process between maven's JVM and Codeserver's JVM anymore .. and after Exit all the JMVs are Stopped. Thx !

**Comment by jjYBdx4IL** [comment]
Easiest solution would be to use jetty's stopPort if possible.

If that's not possible, a cross-platform library to kill processes by port number would be in order. (win10/linux: netstat, win10: tasklist+taskkill, linux; kill)

Here is a quick and dirty solution I used in one of my projects:

https://github.com/jjYBdx4IL/example-maven-project-setups/blob/master/gwt-example/devel-it/src/main/resources/startStop.sh

But it is mixed with Linux where I simply put an environment variable into the started process and then identify the process using that unique tag. That's not possible under Windows it seems.

**Comment by jjYBdx4IL** [comment]
https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4717969

a shutdown hook would at least solve the CTRL-C issue on Windows:

            final ProcessBuilder pb =
                new ProcessBuilder("java", "ControlCProblemDemo");
            Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
                final Process p = pb.start();
                
                @Override
                public void run() {
                    System.out.println("killing it");
                    p.destroyForcibly();
                }
            }));

**Comment by tbroyer** [comment]
There is such a shutdown hook already, with the exception that it calls `destroy()` rather than `destroyForcibly()` (plexus-utils is likely to have/want to support old JVMs though, and `destroyForcibly()`w as added in 1.8).
https://github.com/codehaus-plexus/plexus-utils/blob/plexus-utils-3.1.1/src/main/java/org/codehaus/plexus/util/cli/CommandLineUtils.java#L134-L152
That being said, on Windows, `destroyForcibly()` just calls `destroy()`:
https://github.com/unofficial-openjdk/openjdk/blob/2357b5fc925a96e53e6e9a84230be327424ad1ce/src/java.base/windows/classes/java/lang/ProcessImpl.java#L553-L557

IOW, what you're proposing is already there.

On Windows, a workaround might be to use `gwt:devmode` with `<devmodeArg>-noserver</devmodeArg>`; that way you can easily stop the code server by closing the devmode window.

**Comment by jjYBdx4IL** [comment]
okay, I didn't see that the shutdown hook was already used because the forked plexus-utils repo has a non-working search function. pft.

Apart from that, I played around with plexus-utils to find out why my own test case works, but plexus-utils doesn't. Answer: it seems to be related with the stream pumpers. When I replace Runtime.exec() with ProcessBuilder and use its inheritIO() instead, the child process goes away as expected.

Commandline:

    /**
     * Executes the command.
     */
    public Process execute()
        throws CommandLineException
    {
        // TODO: Provided only for backward compat. with <= 1.4
        verifyShellState();

        Process process;

        // addEnvironment( "MAVEN_TEST_ENVAR", "MAVEN_TEST_ENVAR_VALUE" );

        String[] environment = getEnvironmentVariables();

        File workingDir = shell.getWorkingDirectory();

        ProcessBuilder pb = new ProcessBuilder(getCommandline());
        pb.directory(workingDir);
        pb.inheritIO();
        if (environment != null) {
            for (String s : environment) {
                String[] kv = s.split("=", 2);
                pb.environment().put(kv[0], kv[1]);
            }
        }
        
        try
        {
            if ( workingDir != null ) {
                if ( !workingDir.exists() )
                {
                    throw new CommandLineException( "Working directory \"" + workingDir.getPath()
                        + "\" does not exist!" );
                }
                else if ( !workingDir.isDirectory() )
                {
                    throw new CommandLineException( "Path \"" + workingDir.getPath()
                        + "\" does not specify a directory." );
                }
            }

            process = pb.start();
        }
        catch ( IOException ex )
        {
            throw new CommandLineException( "Error while executing process.", ex );
        }

        return process;
    }

**Comment by jjYBdx4IL** [comment]
the build failure is not because of my changes btw

**Comment by ratuka** [comment]
> You could also probably use `gwt:dev`

Do you mean `gwt:devmode˛`?

**Comment by tbroyer** [comment]
Oops, yes, `gwt:devmode` indeed.

**Comment by ratuka** [comment]
> Oops, yes, `gwt:devmode` indeed.

and without `-noserver`

**Comment by tbroyer** [comment]
**with** `noserver`: devmode is then equivalent to codeserver, just with an additional swing window.

**Comment by craigmit** [comment]
So, if you've created a client/server/shared architect, then instead of `mvn gwt:codeserver -pl *-client -am`, you'd do:
```
mvn gwt:devmode -pl *-client -am -Dnoserver
```
Works great, thanks!  👍

**Comment by FrankHossfeld** [comment]
@tbroyer in case GWT remove the DevMode (as mentioned here: ([https://www.gwtproject.org/release-notes.html#Release_Notes_2_11_0](https://www.gwtproject.org/release-notes.html#Release_Notes_2_11_0)), will this have an impact of this work around?

**Comment by tbroyer** [comment]
For the time being, the goal seems to be to eventually default to `-noserver` (https://github.com/gwtproject/gwt/issues/9863) (and then maybe remove the `JettyLauncher` in favor of the newly added `StaticResourcesServer`, that's at least what I would do if you ask me), but not to remove DevMode entirely.

**Comment by FrankHossfeld** [comment]
Yes, that makes sense. So, the workaround will also work in the future. I was unsure, if it might fail in the future and a new workaround is needed.
Thanks.

**Comment by craigmit** [comment]
> So, if you've created a client/server/shared architect, then instead of `mvn gwt:codeserver -pl *-client -am`, you'd do:
> 
> ```
> mvn gwt:devmode -pl *-client -am -Dnoserver
> ```
> 
> Works great, thanks! 👍

**Note:** After a `mvn clean` the devmode solution will break.  The browser won't be able to find the nocache.js file.

Running the `mvn gwt:codeserver -pl *-client -am` once (and then manually killing it), the devmode solution will start working again.  I'm not entirly sure why, I suspect it's something to do with the target/gwt/codeserver directory it creates.

**Comment by tbroyer** [comment]
Did you configure `warDir` to the same value as codeserver's `launcherDir`? If not, then devmode won't generate the nocache.js where you expect them, that's why you'd need to run codeserver to generate them first.

**Comment by craigmit** [comment]
Ah, thank @tbroyer .  Yes, that was the issue.  Now with:
```
<plugin>
	<groupId>net.ltgt.gwt.maven</groupId>
	<artifactId>gwt-maven-plugin</artifactId>
	<inherited>false</inherited>
	<configuration>
		<launcherDir>${basedir}/drift-team-server/target/classes/launcherDir/</launcherDir>
		<warDir>${basedir}/drift-team-server/target/classes/launcherDir/</warDir>
	</configuration>
</plugin>
```
And still works great after a `mvn clean`.

**Comment by tbroyer** [comment]
Fwiw, I just published version 2024.2.11 of the [gwt-maven-archetypes](https://github.com/tbroyer/gwt-maven-archetypes) that adds the `warDir` configuration, and a note in the README explaining when one may want to use devmode, and linking here.
It also adds the `-noserver` argument.
